design/thin-edge-actors-design.md

```diff
@@ -0,0 +1,510 @@
+    // update self.state
+    self.state += 1;
+
+    // interact with self.peer, without waiting for completion
+    self.peer.say("that").await;
+  }
+}
+```
+
+These three variants of the same example highlight several points.
+
+* The state is managed the same way in the three cases.
+* What differs is the interaction with peers.
+  * The more salient difference is related to the messages.
+  * However, the main difference is the concurrency model.
+    * With `Arc::<Mutex<T>>`, the caller awaits the peer to finalize the request;
+      while with messages the requests are processed concurrently by the peer without blocking the caller.
+    * In the context of thin-edge, concurrent processing is preferred.
+      However, we need to find a way to let the caller wait for response when appropriate.
+* The message passing approach leads to boilerplate code.
+  However, the heart of the code is free of any message construction and deconstruction
+  (see the methods `do_this()` and `do_that()` of the third variant).
+  Furthermore, it seems feasible to generate this boilerplate code using macros
+  (the `run()` method, the handler `ActorAHandler` struct and implementation, the `AMessage` enum).
+* The handler type is a great place to improve both code readability and code flexibility
+  (A wrapping similar to what has been done for the third variant can be done for the first to hide lock acquisition).
+  We will use such handlers and adapters to encapsulate messages related features
+  as sending high priority messages or sending a request for which a response is awaited.
+* Similarly, the channel receiver can be encapsulated to add features
+  like reading high priority messages first or awaiting a response
+* Key points are not addressed by this example.
+  * Who creates the actor channels and instances?
+  * How actors and peers are connected to each others?
+  * Who spawns the actor `run()` method?
+
+
+
+### Messages
+
+Messages must flow freely between actors with no constraints on ownership and thread.
+As they are used to improved observability, they must be `Display`.
+
+```rust
+/// A message exchanged between two actors
+pub trait Message: Clone + Display + Send + Sync + 'static {}
+
+/// There is no need to tag messages as such
+impl<T: Clone + Display + Send + Sync + 'static> Message for T {}
+```
+
+Typical examples of thin-edge messages are telemetry data, operation requests and outcomes,
+but also low level messages as MQTT messages, HTTP requests and responses,
+and even system specific messages as file-system events and update requests.
+
+To be discussed:
+* Do the messages need to be `Clone`?
+  This has to be considered along the idea of using `oneshot` channel for the response to a request.
+  It might be better to be explicit
+  and use `Message + Clone` in contexts where messages are broadcast.
+
+### Channels
+
+Multi-producer single-consumer (`mpsc`) channels are used to exchange messages between actors.
```

Comment:
Since we have a lot of inherent pub-sub interactions, wouldn't we need a multi-producer multi-consumer channel as well to emulate the MQTT broker?

Reply:
This is what we used for the very first version of thin-edge. But it was a mistake because if a consumer is slow than all the others are slowed down. The broadcast channel with used was even discarding messages to cope with such cases. If broadcast is required we can always clone messages over several `mpsc` senders.